File: Data Structures & Algorithms/depth-of-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # Return 0 if the root is None
        if not root:
            return 0

        result = 0

        # Create a stack to store a tuple of node and depth
        # and push the root node with depth 1 into the stack
        s = []
        s.append((root, 1))

        # Exit the while loop when the stack is empty
        while s:
            # Get the current node and depth from the stack
            curr = s.pop()
            (node, depth) = curr

            # If that node exists,
            #   1. Push that node's left/right children and incremented depth into the stack
            #   2. Calculate the max depth
            if node:
                s.append((node.left, depth + 1))
                s.append((node.right, depth + 1))
                result = max(result, depth)
        return result

File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    bool isValidBST(TreeNode* root) {
        // Return true if root is empty
        if (root == nullptr) {
            return true;
        }

        // Create a queue that carries current node and left/right bounds
        // and push root into the queue so we can start the while loop
        queue<tuple<TreeNode*, long long, long long>> q;
        q.push(make_tuple(root, LLONG_MIN, LLONG_MAX));

        // Exit the while loop when queue is empty
        while (!q.empty()) {
            // Get the front node and remove it from the queue
            auto [node, left, right] = q.front();
            q.pop();

            // Continue is current node is empty because we don't need to process it
            if (node == nullptr) {
                continue;
            }

            // Return false if current node is not valid for any of left/right bounds
            if (!(node->val > left && node->val < right)) {
                return false;
            }

            q.push(make_tuple(node->left, left, node->val));
            q.push(make_tuple(node->right, node->val, right));
        }

        // Return true if we've successfully processed all nodes in the BST
        return true;
    }
};
